### PR TITLE
add io.github.NickKarpowicz.LightwaveExplorer

### PR DESCRIPTION
**Pull request by NickKarpowicz**
This is an open source nonlinear optics simulation that is both easy to use and powerful. It has already been used for several publications, and has an emphasis on visual feedback.

### CODE REVIEW

**Comment by hfiguiere** [comment]
the PR template include a set of question that you have erased.

File: .github/pull_request_template.md
Comment (hfiguiere) [inline .github/pull_request_template.md:1]: the content of the `.github` directory shouldn't be part of this PR.

File: CONTRIBUTING.md
Comment (hfiguiere) [inline CONTRIBUTING.md:1]: this doesn't belong here either.

File: COPYING
Comment (hfiguiere) [inline COPYING:1]: this doesn't belong here either.

File: README.md
Comment (hfiguiere) [inline README.md:1]: this doesn't belong here either.

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            "version": "22.08",
+            "add-ld-path": "."
+        }
+    },
+    "command": "flatpakLauncher.sh",
+    "finish-args": [
+        "--share=ipc",
+        "--device=dri",
+        "--socket=x11",
+        "--socket=wayland",
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:21]: if it supports waylandm then line 20 should have `--fallback-x11`

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        "--filesystem=host"
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:22]: `host` is a very broad file system permission

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                            "rm l_BaseKit_p_2023.2.0.49397_offline.sh",
+                            "webimage_extracted/l_BaseKit_p_2023.2.0.49397_offline/bootstrapper -s --action=install --components=intel.oneapi.lin.mkl.devel --install-dir=/app/opt/intel/oneapi --eula=accept",
+                            "rm -rf /app/opt/intel/oneapi/mkl/latest/lib/ia32",
+                            "rm -f /app/opt/intel/oneapi/mkl/latest/lib/intel64/*.so*",
+                            "rm -rf /app/opt/intel/oneapi/conda_channel"],
+            "sources": [
+                {
+                    "type": "file",
+                    "url": "https://registrationcenter-download.intel.com/akdlm/IRC_NAS/992857b9-624c-45de-9701-f6445d845359/l_BaseKit_p_2023.2.0.49397_offline.sh",
+                    "md5": "560466f714a79fce3cb8793528d487c8"
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:85]: `md5` is deprecated. Please use `sha1`

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        
+                }
+            ]
+        },
+        {
+            "name": "fmt",
+            "config-opts": ["-DCMAKE_POSITION_INDEPENDENT_CODE=TRUE", "-DFMT_TEST=OFF"],
+            "buildsystem": "cmake",
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:93]:
```suggestion
            "buildsystem": "cmake-ninja",
```

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+                { 
+                    "type": "git", 
+                    "url": "https://github.com/fmtlib/fmt", 
+                    "tag": "10.1.0",
+                    "commit": "e57ca2e3685b160617d3d95fcd9e789c4e06ca88" } 
+                ]
+        },
+        {
+            "name": "gcem",
+            "buildsystem": "cmake",
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:104]:
```suggestion
            "buildsystem": "cmake-ninja",
```

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                    "type": "git", 
+                    "url": "https://github.com/davisking/dlib", 
+                    "tag": "v19.24.2",
+                    "commit": "39749fba5163d4e25030358f6aa8d91617bfdca2" } 
+                ]
+        },
+        {
+            "name": "cuda",
+            "buildsystem": "simple",
+            "build-commands":["mkdir /app/tmp","sh cuda_12.2.1_535.86.10_linux.run --silent --toolkitpath=/app/opt/cuda --toolkit --override --tmpdir=/app/tmp"],
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:128]: how does this work on `aarch64`?

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+            "echo \". /app/opt/intel/oneapi/setvars.sh\" >> localscript.sh",
+            "echo \"export PATH=$PATH:/app/opt/cuda/bin\" >> localscript.sh",
+            "echo \"export CMAKE_PREFIX_PATH=$CMAKE_PREFIX_PATH:/app/opt/intel/oneapi/mkl/latest/lib/cmake/mkl\" >> localscript.sh",
+            "echo \"export CUDA_PATH=$PATH:/app/opt/cuda/\" >> localscript.sh",
+            "echo \"cmake -DMAKECUDAFLATPAK=TRUE -DCMAKE_CUDA_ARCHITECTURES=75 .\" >> localscript.sh",
+            "echo \"make\" >> localscript.sh",
+            "chmod +x localscript.sh","./localscript.sh",
+            "cmake --install . --prefix \"/app\""],
+            "build-options": {
+                "cxxflags": "-O3 -g -pipe -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong -grecord-gcc-switches",
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:150]: do you really need these flags? Most are set by default.

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+                "cxxflags-override": true,
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:151]: on general this is not recommended.

**Review by hfiguiere — COMMENTED** [review]
I do not see any desktop, icon nor appstream file installed.

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+            "buildsystem": "simple",
+            "build-commands": ["touch localscript.sh",
+            "echo \". /app/opt/intel/oneapi/setvars.sh\" >> localscript.sh",
+            "echo \"export PATH=$PATH:/app/opt/cuda/bin\" >> localscript.sh",
+            "echo \"export CMAKE_PREFIX_PATH=$CMAKE_PREFIX_PATH:/app/opt/intel/oneapi/mkl/latest/lib/cmake/mkl\" >> localscript.sh",
+            "echo \"export CUDA_PATH=$PATH:/app/opt/cuda/\" >> localscript.sh",
+            "echo \"cmake -DMAKECUDAFLATPAK=TRUE -DCMAKE_CUDA_ARCHITECTURES=75 .\" >> localscript.sh",
+            "echo \"make\" >> localscript.sh",
+            "chmod +x localscript.sh","./localscript.sh",
+            "cmake --install . --prefix \"/app\""],
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:148]: this should be buildsystem `cmake-ninja`

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        },
+        {
+            "name": "cuda",
+            "buildsystem": "simple",
+            "build-commands":["mkdir /app/tmp","sh cuda_12.2.1_535.86.10_linux.run --silent --toolkitpath=/app/opt/cuda --toolkit --override --tmpdir=/app/tmp"],
+            "sources":[
+                {
+                    "type": "file",
+                    "url": "https://developer.download.nvidia.com/compute/cuda/12.2.1/local_installers/cuda_12.2.1_535.86.10_linux.run",
+                    "md5": "2579fcb61cd80d3284d73da93785550f"
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:133]: use `sha256`

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 11]
```diff
@@ -0,0 +1,165 @@
+{
+    "id": "io.github.NickKarpowicz.LightwaveExplorer",
+    "branch": "master",
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:3]:
```suggestion
```

never set branch `master`.

File: io.github.NickKarpowicz.LightwaveExplorer.json [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    "runtime": "org.gnome.Platform",
```
Comment (hfiguiere) [inline io.github.NickKarpowicz.LightwaveExplorer.json:4]: is the GNOME runtime really necessary?

**Comment by hfiguiere** [comment]
> I do not see any desktop, icon nor appstream file installed.

I missed that `cmake`. Which is why it should be `buildsystem: cmake-ninja`.

File: .github/pull_request_template.md
Comment (NickKarpowicz) [reply .github/pull_request_template.md:1]: Apologies for the mess - I didn't mean to include the stuff from the master branch. Re: the list of checks, I must have wiped that out by submitting the PR via the Github-CLI interface... I'll put them in a comment below if that's ok, alternatively I could make a new PR to fix these issues?

File: io.github.NickKarpowicz.LightwaveExplorer.json (hunk 1 above)
Comment (NickKarpowicz) [reply io.github.NickKarpowicz.LightwaveExplorer.json:21]: good to know, I'll add it.

File: io.github.NickKarpowicz.LightwaveExplorer.json (hunk 2 above)
Comment (NickKarpowicz) [reply io.github.NickKarpowicz.LightwaveExplorer.json:22]: It's unfortunately necessary because of the pattern of saving: it has been quite common for students to queue up a batch of simulations that will take a day to complete. As far as I can tell, the portal that I get through GTK cannot handle this - I just take a path from the user and write there ASAP to preserve data...

File: io.github.NickKarpowicz.LightwaveExplorer.json (hunk 6 above)
Comment (NickKarpowicz) [reply io.github.NickKarpowicz.LightwaveExplorer.json:128]: It could in principle work by using the aarch64 runfile, but the bigger problem is with OneAPI which is required as well, which would not support the platform. I would only build for x86_64. On the Mac version, I replace the OneAPI FFT routine with FFTW such that aarch64 can be built for, but FFTW (GPL3.0) is not license-compatible (afaik) with CUDA (which isn't available on Mac anymore anyway).

File: io.github.NickKarpowicz.LightwaveExplorer.json (hunk 7 above)
Comment (NickKarpowicz) [reply io.github.NickKarpowicz.LightwaveExplorer.json:150]: These are essentially the default flags, but I need to remove D_GLIBCXX_ASSERTIONS, which throws an exception when a std::complex NaN is encountered, before the simulation's own error-handling deals with it. This causes the program to throw a not-very-informative assertion failure rather than an in-GUI message giving hints at how to fix the input parameters.

File: io.github.NickKarpowicz.LightwaveExplorer.json (hunk 12 above)
Comment (NickKarpowicz) [reply io.github.NickKarpowicz.LightwaveExplorer.json:4]: The GUI is based on GTK4, which is provided by the gnome runtime

**Comment by NickKarpowicz** [comment]
Since there were a number of problems with this request, I am going to close it and make a new one, with the new manifest I just tested which incorporates the above comments.

**Comment by hfiguiere** [comment]
Closing the PR insteead of fixing it waste the time of the reviewers.
